File: backend/app/services/kyc_service.py
from datetime import datetime, timezone
from typing import Dict, Any, List, Optional
from bson import ObjectId
from pymongo.database import Database
from fastapi import HTTPException, status

from app.schemas.kyc import (
    KYCSubmitRequest,
    KYCResponse,
    AdminKYCPendingItem,
    AdminKYCPendingListResponse,
    AdminKYCUserProfile,
    AdminKYCDetailResponse,
    KYCActionResponse,
    AddressSchema,
)

_kyc_indexes_initialized = False


def ensure_kyc_indexes(db: Database):
    """Ensure unique index on user_id in kyc collection."""
    global _kyc_indexes_initialized
    if _kyc_indexes_initialized or db is None:
        return
    try:
        db.kyc.create_index("user_id", unique=True, name="uniq_kyc_user_id")
        _kyc_indexes_initialized = True
    except Exception:
        pass


def _to_object_id(id_val: Any) -> Optional[ObjectId]:
    """Helper to safely convert string to ObjectId."""
    if isinstance(id_val, ObjectId):
        return id_val
    try:
        return ObjectId(str(id_val))
    except Exception:
        return None


def format_kyc_response(doc: Dict[str, Any]) -> KYCResponse:
    """Format raw MongoDB KYC document to KYCResponse schema."""
    addr_data = doc.get("address", {})
    if isinstance(addr_data, dict):
        address = AddressSchema(
            address_line=addr_data.get("address_line", ""),
            city=addr_data.get("city", ""),
            state=addr_data.get("state", ""),
            pincode=addr_data.get("pincode", ""),
        )
    else:
        address = AddressSchema(address_line="", city="", state="", pincode="")

    return KYCResponse(
        id=str(doc["_id"]),
        user_id=str(doc["user_id"]),
        full_name=doc.get("full_name", ""),
        date_of_birth=str(doc.get("date_of_birth", "")),
        gender=doc.get("gender", ""),
        address=address,
        id_type=doc.get("id_type", ""),
        id_number=doc.get("id_number", ""),
        status=doc.get("status", "pending"),
        rejection_reason=doc.get("rejection_reason"),
        submitted_at=doc.get("submitted_at"),
        reviewed_at=doc.get("reviewed_at"),
        reviewed_by=str(doc["reviewed_by"]) if doc.get("reviewed_by") else None,
        updated_at=doc.get("updated_at"),
    )


def submit_kyc(db: Database, current_user: Dict[str, Any], data: KYCSubmitRequest) -> KYCResponse:
    """Submit customer KYC details or resubmit a previously rejected KYC."""
    if db is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Database service unavailable",
        )

    ensure_kyc_indexes(db)
    user_id_obj = _to_object_id(current_user["id"])
    if not user_id_obj:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid user ID")

    existing_kyc = db.kyc.find_one({"user_id": user_id_obj})
    now = datetime.now(timezone.utc)

    if existing_kyc:
        existing_status = existing_kyc.get("status", "pending")
        if existing_status == "pending":
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="KYC verification is already pending",
            )
        if existing_status == "verified":
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="KYC is already verified",
            )
        
        # If status is rejected, update the existing record
        update_fields = {
            "full_name": data.full_name,
            "date_of_birth": data.date_of_birth,
            "gender": data.gender,
            "address": data.address.model_dump(),
            "id_type": data.id_type,
            "id_number": data.id_number,
            "status": "pending",
            "rejection_reason": None,
            "submitted_at": now,
            "reviewed_at": None,
            "reviewed_by": None,
            "updated_at": now,
        }
        db.kyc.update_one({"_id": existing_kyc["_id"]}, {"$set": update_fields})
        db.users.update_one({"_id": user_id_obj}, {"$set": {"kyc_status": "pending", "updated_at": now}})
        
        updated_doc = db.kyc.find_one({"_id": existing_kyc["_id"]})
        return format_kyc_response(updated_doc)

    # New KYC submission
    new_kyc_doc = {
        "user_id": user_id_obj,
        "full_name": data.full_name,
        "date_of_birth": data.date_of_birth,
        "gender": data.gender,
        "address": data.address.model_dump(),
        "id_type": data.id_type,
        "id_number": data.id_number,
        "status": "pending",
        "rejection_reason": None,
        "submitted_at": now,
        "reviewed_at": None,
        "reviewed_by": None,
        "updated_at": now,
    }

    result = db.kyc.insert_one(new_kyc_doc)
    db.users.update_one({"_id": user_id_obj}, {"$set": {"kyc_status": "pending", "updated_at": now}})
    new_kyc_doc["_id"] = result.inserted_id
    return format_kyc_response(new_kyc_doc)


def get_user_kyc(db: Database, current_user: Dict[str, Any]) -> KYCResponse:
    """Retrieve current customer's KYC details."""
    if db is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Database service unavailable",
        )

    user_id_obj = _to_object_id(current_user["id"])
    kyc_doc = db.kyc.find_one({"user_id": user_id_obj})
    if not kyc_doc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="KYC record not found",
        )
    return format_kyc_response(kyc_doc)


def get_pending_kyc_list(db: Database) -> AdminKYCPendingListResponse:
    """Retrieve all pending KYC records for Admin review."""
    if db is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Database service unavailable",
        )

    cursor = db.kyc.find({"status": "pending"}).sort("submitted_at", -1)
    items: List[AdminKYCPendingItem] = []

    for kyc_doc in cursor:
        user_id_obj = kyc_doc.get("user_id")
        user_doc = db.users.find_one({"_id": user_id_obj}) if user_id_obj else None
        
        user_name = user_doc.get("name") if user_doc else kyc_doc.get("full_name", "")
        user_mobile = user_doc.get("mobile") if user_doc else ""

        items.append(
            AdminKYCPendingItem(
                kyc_id=str(kyc_doc["_id"]),
                user_id=str(user_id_obj),
                name=user_name,
                mobile=user_mobile,
                status="pending",
                submitted_at=kyc_doc.get("submitted_at"),
            )
        )

    return AdminKYCPendingListResponse(items=items, total=len(items))


def get_kyc_details(db: Database, kyc_id: str) -> AdminKYCDetailResponse:
    """Retrieve full KYC details along with associated user profile for Admin."""
    if db is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Database service unavailable",
        )

    kyc_obj_id = _to_object_id(kyc_id)
    if not kyc_obj_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid KYC ID",
        )

    kyc_doc = db.kyc.find_one({"_id": kyc_obj_id})
    if not kyc_doc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="KYC record not found",
        )

    user_id_obj = kyc_doc.get("user_id")
    user_doc = db.users.find_one({"_id": user_id_obj})
    if not user_doc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Associated user not found",
        )

    user_profile = AdminKYCUserProfile(
        id=str(user_doc["_id"]),
        name=user_doc.get("name", ""),
        mobile=user_doc.get("mobile", ""),
        email=user_doc.get("email"),
        account_status=user_doc.get("account_status", "active"),
        kyc_status=user_doc.get("kyc_status", "pending"),
    )

    return AdminKYCDetailResponse(
        kyc=format_kyc_response(kyc_doc),
        user=user_profile,
    )


def approve_kyc(db: Database, kyc_id: str, admin_user: Dict[str, Any]) -> KYCActionResponse:
    """Admin action: Approve customer KYC verification."""
    if db is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Database service unavailable",
        )

    kyc_obj_id = _to_object_id(kyc_id)
    if not kyc_obj_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid KYC ID",
        )

    kyc_doc = db.kyc.find_one({"_id": kyc_obj_id})
    if not kyc_doc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="KYC record not found",
        )

    current_status = kyc_doc.get("status")
    if current_status == "verified":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="KYC is already verified",
        )

    now = datetime.now(timezone.utc)
    admin_id = str(admin_user.get("id", admin_user.get("_id")))

    # Update KYC document
    db.kyc.update_one(
        {"_id": kyc_obj_id},
        {
            "$set": {
                "status": "verified",
                "rejection_reason": None,
                "reviewed_at": now,
                "reviewed_by": admin_id,
                "updated_at": now,
            }
        }
    )

    # Synchronize users collection
    user_id_obj = kyc_doc.get("user_id")
    if user_id_obj:
        db.users.update_one(
            {"_id": user_id_obj},
            {"$set": {"kyc_status": "verified", "updated_at": now}}
        )

    return KYCActionResponse(
        message="KYC verified successfully",
        status="verified",
    )


def reject_kyc(db: Database, kyc_id: str, admin_user: Dict[str, Any], reason: str) -> KYCActionResponse:
    """Admin action: Reject customer KYC verification with mandatory reason."""
    if db is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Database service unavailable",
        )

    kyc_obj_id = _to_object_id(kyc_id)
    if not kyc_obj_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid KYC ID",
        )

    kyc_doc = db.kyc.find_one({"_id": kyc_obj_id})
    if not kyc_doc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="KYC record not found",
        )

    current_status = kyc_doc.get("status")
    if current_status == "verified":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Cannot reject an already verified KYC",
        )

    now = datetime.now(timezone.utc)
    admin_id = str(admin_user.get("id", admin_user.get("_id")))

    # Update KYC document
    db.kyc.update_one(
        {"_id": kyc_obj_id},
        {
            "$set": {
                "status": "rejected",
                "rejection_reason": reason,
                "reviewed_at": now,
                "reviewed_by": admin_id,
                "updated_at": now,
            }
        }
    )

    # Synchronize users collection
    user_id_obj = kyc_doc.get("user_id")
    if user_id_obj:
        db.users.update_one(
            {"_id": user_id_obj},
            {"$set": {"kyc_status": "rejected", "updated_at": now}}
        )

    return KYCActionResponse(
        message="KYC rejected",
        status="rejected",
    )

File: backend/src/services/kycService.js
import { v4 as uuidv4 } from 'uuid';
import { query } from '../config/db.js';
import { notifyKycSubmitted, notifyKycApproved, notifyKycRejected } from './notificationService.js';

export function formatKycResponse(doc) {
  return {
    id: doc.id,
    user_id: doc.user_id,
    full_name: doc.full_name,
    date_of_birth: doc.date_of_birth,
    gender: doc.gender,
    address: {
      address_line: doc.address_line || '',
      city: doc.city || '',
      state: doc.state || '',
      pincode: doc.pincode || '',
    },
    id_type: doc.id_type,
    id_number: doc.id_number,
    status: doc.status,
    rejection_reason: doc.rejection_reason || null,
    submitted_at: doc.submitted_at,
    reviewed_at: doc.reviewed_at || null,
    reviewed_by: doc.reviewed_by || null,
    updated_at: doc.updated_at,
  };
}

export async function submitKyc(user, data) {
  const existingRows = await query('SELECT * FROM kyc WHERE user_id = ? LIMIT 1', [user.id]);

  if (existingRows.length > 0) {
    const existing = existingRows[0];
    if (existing.status === 'pending') {
      const error = new Error('KYC verification is already pending');
      error.status = 400;
      throw error;
    }
    if (existing.status === 'verified') {
      const error = new Error('KYC is already verified');
      error.status = 400;
      throw error;
    }

    // Re-submission after rejection
    await query(
      `UPDATE kyc SET 
        full_name = ?,
        date_of_birth = ?,
        gender = ?,
        address_line = ?,
        city = ?,
        state = ?,
        pincode = ?,
        id_type = ?,
        id_number = ?,
        status = 'pending',
        rejection_reason = NULL,
        submitted_at = NOW(),
        reviewed_at = NULL,
        reviewed_by = NULL,
        updated_at = NOW()
       WHERE id = ?`,
      [
        data.full_name,
        data.date_of_birth,
        data.gender,
        data.address?.address_line || '',
        data.address?.city || '',
        data.address?.state || '',
        data.address?.pincode || '',
        data.id_type,
        data.id_number,
        existing.id,
      ]
    );

    await query('UPDATE users SET kyc_status = \'pending\', updated_at = NOW() WHERE id = ?', [user.id]);
    const updated = await query('SELECT * FROM kyc WHERE id = ?', [existing.id]);
    return formatKycResponse(updated[0]);
  }

  // New KYC Submission
  const kycId = uuidv4();
  await query(
    `INSERT INTO kyc 
      (id, user_id, full_name, date_of_birth, gender, address_line, city, state, pincode, id_type, id_number, status, rejection_reason, submitted_at, updated_at)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'pending', NULL, NOW(), NOW())`,
    [
      kycId,
      user.id,
      data.full_name,
      data.date_of_birth,
      data.gender,
      data.address?.address_line || '',
      data.address?.city || '',
      data.address?.state || '',
      data.address?.pincode || '',
      data.id_type,
      data.id_number,
    ]
  );

  await query('UPDATE users SET kyc_status = \'pending\', updated_at = NOW() WHERE id = ?', [user.id]);

  try {
    await notifyKycSubmitted(user.id, kycId);
  } catch (err) {}

  const created = await query('SELECT * FROM kyc WHERE id = ?', [kycId]);
  return formatKycResponse(created[0]);
}

export async function getUserKyc(user) {
  const rows = await query('SELECT * FROM kyc WHERE user_id = ? LIMIT 1', [user.id]);
  if (rows.length === 0) {
    const error = new Error('KYC record not found');
    error.status = 404;
    throw error;
  }
  return formatKycResponse(rows[0]);
}

export async function getPendingKycList() {
  const rows = await query(
    `SELECT k.id as kyc_id, k.user_id, k.full_name, k.status, k.submitted_at, u.name as user_name, u.mobile as user_mobile 
     FROM kyc k
     JOIN users u ON k.user_id = u.id
     WHERE k.status = 'pending'
     ORDER BY k.submitted_at DESC`
  );

  const items = rows.map((r) => ({
    kyc_id: r.kyc_id,
    user_id: r.user_id,
    name: r.user_name || r.full_name,
    mobile: r.user_mobile,
    status: 'pending',
    submitted_at: r.submitted_at,
  }));

  return {
    items,
    total: items.length,
  };
}

export async function getKycDetails(kycId) {
  const rows = await query('SELECT * FROM kyc WHERE id = ? LIMIT 1', [kycId]);
  if (rows.length === 0) {
    const error = new Error('KYC record not found');
    error.status = 404;
    throw error;
  }
  const kyc = rows[0];

  const userRows = await query('SELECT id, name, mobile, email, account_status, kyc_status FROM users WHERE id = ? LIMIT 1', [kyc.user_id]);
  if (userRows.length === 0) {
    const error = new Error('Associated user not found');
    error.status = 404;
    throw error;
  }

  return {
    kyc: formatKycResponse(kyc),
    user: userRows[0],
  };
}

export async function approveKyc(kycId, adminUser) {
  const rows = await query('SELECT * FROM kyc WHERE id = ? LIMIT 1', [kycId]);
  if (rows.length === 0) {
    const error = new Error('KYC record not found');
    error.status = 404;
    throw error;
  }

  const kyc = rows[0];
  if (kyc.status === 'verified') {
    const error = new Error('KYC is already verified');
    error.status = 400;
    throw error;
  }

  await query(
    `UPDATE kyc SET status = 'verified', rejection_reason = NULL, reviewed_at = NOW(), reviewed_by = ?, updated_at = NOW() WHERE id = ?`,
    [adminUser.id, kycId]
  );
  await query('UPDATE users SET kyc_status = \'verified\', updated_at = NOW() WHERE id = ?', [kyc.user_id]);

  try {
    await notifyKycApproved(kyc.user_id, kycId);
  } catch (err) {}

  return {
    message: 'KYC verified successfully',
    status: 'verified',
  };
}

export async function rejectKyc(kycId, adminUser, reason) {
  const rows = await query('SELECT * FROM kyc WHERE id = ? LIMIT 1', [kycId]);
  if (rows.length === 0) {
    const error = new Error('KYC record not found');
    error.status = 404;
    throw error;
  }

  const kyc = rows[0];
  if (kyc.status === 'verified') {
    const error = new Error('Cannot reject an already verified KYC');
    error.status = 400;
    throw error;
  }

  await query(
    `UPDATE kyc SET status = 'rejected', rejection_reason = ?, reviewed_at = NOW(), reviewed_by = ?, updated_at = NOW() WHERE id = ?`,
    [reason, adminUser.id, kycId]
  );
  await query('UPDATE users SET kyc_status = \'rejected\', updated_at = NOW() WHERE id = ?', [kyc.user_id]);

  try {
    await notifyKycRejected(kyc.user_id, kycId, reason);
  } catch (err) {}

  return {
    message: 'KYC rejected',
    status: 'rejected',
  };
}

export default {
  submitKyc,
  getUserKyc,
  getPendingKycList,
  getKycDetails,
  approveKyc,
  rejectKyc,
};
